File: 프로그래머스/lv2/92341. 주차 요금 계산/주차 요금 계산.py
from math import ceil


def solution(fees, records):
    basicTime, basicFee, timeUnit, unitPerFee = fees
    parkDict = dict()
    for record in records:
        time, serial, content = record.split()
        hour, minute = map(int, time.split(':'))
        if len(content) < 3:
            if serial in parkDict:
                parkDict[serial][1] = hour * 60 + minute
            else:
                parkDict[serial] = [0, hour * 60 + minute]
        else:
            parkDict[serial][0] += hour * 60 + minute - parkDict[serial][1]
            parkDict[serial][1] = -1
    
    for serial, [totalTime, lastTime] in parkDict.items():
        parkDict[serial] = basicFee
        if lastTime > -1:
            totalTime += 24 * 60 - lastTime - 1
        if totalTime > basicTime:
            parkDict[serial] += ceil((totalTime - basicTime) / timeUnit) * unitPerFee
    return [value for _, value in sorted(parkDict.items())]

File: 프로그래머스/lv2/92341. 주차 요금 계산/주차 요금 계산.kt
import java.util.TreeMap;
import java.util.StringTokenizer;

class Solution {

    val MAX_TIME = 24 * 60 - 1
    val totalTimes = TreeMap<Int, Int>()
    val inRecords = HashMap<Int, Int>()
    var basicPer = 0
    var basicFee = 0
    var extraPer = 0
    var extraFee = 0
    
    fun solution(fees: IntArray, records: Array<String>): IntArray {
        basicPer = fees[0]
        basicFee = fees[1]
        extraPer = fees[2]
        extraFee = fees[3]
        
        for (s in records) {
            val (time, number, info) = parse(s)
            
            if (info == "IN") {
                inRecords[number] = time
                continue
            }
            
            val diff = time - inRecords[number]!!
                
            totalTimes[number] = (totalTimes[number] ?: 0) + diff
            inRecords -= number
        }
        
        for ((number, time) in inRecords) {
            val diff = MAX_TIME - time
                
            totalTimes[number] = (totalTimes[number] ?: 0) + diff
        }
        
        val res = IntArray(totalTimes.size)
        var i = 0
        
        for (time in totalTimes.values) {
            res[i++] = calcFee(time)
        }
        
        return res
    }
    
    fun parse(s: String): Record {
        val st = StringTokenizer(s)
        val (h, m) = st.nextToken().split(':')
        val time = h.toInt() * 60 + m.toInt()
        val number = st.nextToken().toInt()
        val info = st.nextToken()
        
        return Record(time, number, info)
    }
    
    fun calcFee(totalTime: Int): Int {
        if (totalTime <= basicPer) {
            return basicFee
        }
        
        var ret = basicFee
        val diff = totalTime - basicPer
        
        ret += extraFee * (diff / extraPer)
        
        return if (diff % extraPer > 0) 
            ret + extraFee
        else 
            ret
    }
}

data class Record(val time: Int, val number: Int, val info: String)